runtime/ms-rest-azure/lib/login.js

```diff
@@ -468,4 +470,232 @@ exports.withServicePrincipalSecretWithAuthResponse = function withServicePrincip
+ */
+function _withAuthFile(options, callback) {
+  if (!callback && typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  if (!options) options = { filePath: '' };
+  if (!callback) {
+    throw new Error('callback cannot be null or undefined.');
+  }
+
+  let filePath = options.filePath;
+  let authLocation = process.env[azureConstants.AZURE_AUTH_LOCATION];
+  let subscriptionEnvVariableName = options.subscriptionEnvVariableName || 'AZURE_SUBSCRIPTION_ID';
+  if (!filePath) {
+    if (!authLocation) {
+      let msg = `Either provide an absolute file path to the auth file or set/export the environment variable - ${azureConstants.AZURE_AUTH_LOCATION}.`;
+      return callback(new Error(msg));
+    }
+    filePath = authLocation;
+  }
+  //expand ~ to user's home directory.
+  if(filePath.startsWith('~')) {
+    filePath = msRest.homeDir(filePath.slice(1));
+  }
+
+  let content = null, credsObj = {}, optionsForSpSecret = {};
+  try {
+    content = fs.readFileSync(filePath, {encoding: 'utf8'});
+    credsObj = JSON.parse(content);
+    _validateAuthFileContent(credsObj, filePath);
+    if (!process.env[subscriptionEnvVariableName]) {
+      process.env[subscriptionEnvVariableName] = credsObj.subscriptionId;
+    }
+  } catch(err) {
+    return callback(err);
+  }
+
+  //get the AzureEnvironment or create a new AzureEnvironment based on the info provided in the auth file
+  let envFound = { 
+    state: false,
+    name: ''
+  };
+  let envNames = Object.keys(Object.getPrototypeOf(AzureEnvironment)).slice(1);
+  for(let i = 0; i < envNames.length; i++) {
+    let env = envNames[i];
+    let environmentObj = AzureEnvironment[env];
+    if (environmentObj && 
+      environmentObj.managementEndpointUrl && 
+      _foundManagementEndpointUrl(credsObj.managementEndpointUrl, environmentObj.managementEndpointUrl)) {
+      envFound.state = true;
+      envFound.name = environmentObj.name;
+      break;
+    }
+  }
+  if (envFound.state) {
+    optionsForSpSecret.environment = AzureEnvironment[envFound.name];
+  } else {
+    //create a new environment with provided info.
+    let envParams = {
+      //try to find a logical name or set the filepath as the env name.
+      name: credsObj.managementEndpointUrl.match(/.*management\.core\.(.*)\..*/i)[1] || filePath
+    };
+    let keys = Object.keys(credsObj);
+    for (let i = 0; i < keys.length; i++) {
+      let key = keys[i];
+      if (key.match(/^(clientId|clientSecret|subscriptionId|tenantId)$/ig) === null) {
+        if (key === 'activeDirectoryEndpointUrl' && !key.endsWith('/')) {
+          envParams[key] = credsObj[key] + '/';
+        } else {
+          envParams[key] = credsObj[key];
+        }
+      }
+    }
+    if (!envParams.activeDirectoryResourceId) {
+      envParams.activeDirectoryResourceId = credsObj.managementEndpointUrl;
+    }
+    if (!envParams.portalUrl) {
+      envParams.portalUrl = 'https://portal.azure.com';
```

Comment:
This is not an environment aware url. If SDK never uses it, which I believe is the case, then maybe we just remove it from the environment